File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        boolean answer = true;
        Stack<Character> stack = new Stack<Character>();

        //'('만 스택에 push하고 ')' 가 들어오면 스택에 (가 있으면 pop으로 한쌍 제거 없으면 false
        for(int i=0 ; i<s.length() ; i++){
            if(s.charAt(i) == '('){
                stack.push(s.charAt(i));
            }
            else{
                if(stack.isEmpty()){
                    return false;
                }
                else{
                    stack.pop();
                }
            }
        }
        answer = (stack.isEmpty() ? true : false);
        return answer;
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.js
function solution(s) {
  const stack = [];

  s.split('').map((char) => {
    if (char === '(') {
      return stack.push(char);
    }

    if (char === ')' && stack[stack.length - 1] === '(') {
      return stack.pop();
    }

    if (char === ')' && stack[stack.length - 1] !== '(') {
      return stack.push(char);
    }
  });

  return stack.length === 0;
}